teps/0138-decouple-api-and-feature-versioning.md

```diff
@@ -72,8 +85,116 @@ Per [TEP-0033](https://github.com/tektoncd/community/blob/main/teps/0033-tekton-
+
+This chart would apply to the existing beta features (array results, array indexing, object params and results, and remote resolution):
+
+| enable-api-fields |	legacy-enable-beta-features-by-default | enabled in v1beta1? | enabled in v1? |
+| ------ | ------ | --- | --- |
+| beta   |	true  |	yes |	yes |
+| beta   |	false |	yes	| yes |
+| stable |	true  |	yes	| no  |
+| stable |	false |	no	| no  |
+
+For new beta features(e.g. matrix in the future):
+
+| enable-api-fields |	legacy-enable-beta-features-by-default | enabled in v1beta1? | enabled in v1? |
+| ------ | ------ | --- | --- |
+| beta   |	true  |	yes |	yes |
+| beta   |	false |	yes |	yes |
+| stable |	true  |	no  |	no  |
+| stable |	false |	no  |	no  |
+
+Once all existing beta features become stable, `legacy-enable-beta-features-by-default` can be removed and we will deprecate and then remove `legacy-enable-beta-features-by-default` and to use `stable` `enable-api-fields`. We would default `true` for the new flag and after 9 months, we default `enable-api-fields` to `stable` to preserve the existing behavior.
+
+#### Pros
+- This will provide a smoother transition for switching default `enable-api-fields` value back to `stable`.
+
+#### Cons
+- It is not clear what to do with alpha features for the feature promotion process when there are two `enable-api-fields` related flags, which might lead to confusions.
+- This is adding complexity to both the implementations and the testing matrix for the newly introduced flag.
+
+### Make `beta` feature validation changes now; migrate `enable-api-fields` to `stable` in 9 months
+Require `enable-api-fields` to be set to `beta` when using beta features, regardless of CRD API version. Immediately make this change for existing beta features. This will solve the unintended behavior that taskruns cannot be created for version coupled features with “enable-api-fields” set to `stable` during the `v1` storage swap right away.
+
+For the default value of `enable-api-fields` in the long run, this alternative will give a nine month notice for users for the breaking change setting the default to `stable`.
+
+#### Pros
+- This solution could provide best clarity of feature stability levels across different apiVersions by keeping validation for beta features the same in `v1` and `v1beta1`.
+
+#### Cons
+- Updating existing beta features valiations in `v1beta1` is a backwards incompatible change. This will break users who are accidentally using the coupled beta features while `enable-api-fields` is set to stable.
+
+### New `legacy-stable` value for `enable-api-fields`; migrate `enable-api-fields` to `stable` in 9 months
+Change the validation for existing default `beta` value of `enable-api-fields` to validate all beta features across apiVersions. Add a `legacy-stable` value to the `enable-api-fields` to keep the behavior of the current "stable" feature validations in `v1beta1` that includes the coupled features. 
```

Comment:
I don't think this solution would involve any changes when enable-api-fields is set to "beta"; instead, it would change the validation of beta features when enable-api-fields is set to "stable".

Reply:
That makes sense, I changed the wording regarding stable.